src/Products/Viewer/Controllers/ViewerApiController.cs

```diff
@@ -383,196 +411,318 @@ public HttpResponseMessage PrintPdf(PostedDataEntity loadDocumentRequest)
-                    allPages.Add(pages[i].HtmlContent);
+                    viewer.View(viewOptions, page.Number);
+
+                    return Encoding.UTF8.GetString(pages[0].ToArray());
                 }
-            }
-            else
-            {
-                ImageOptions imageOptions = new ImageOptions();
-                SetOptions(imageOptions, password, 0);
-                var pages = this.GetHandler().GetPages(documentGuid, imageOptions);
-                for (int i = 0; i < pages.Count; i++)
+                else
                 {
-                    byte[] bytes;
-                    using (var memoryStream = new MemoryStream())
-                    {
-                        pages[i].Stream.CopyTo(memoryStream);
-                        bytes = memoryStream.ToArray();
-                    }
-                    string encodedImage = Convert.ToBase64String(bytes);
-                    allPages.Add(encodedImage);
+                    PngViewOptions pngViewOptions = new PngViewOptions(pageStreamFactory);
+
+                    viewer.View(pngViewOptions, page.Number);
+
+                    return Convert.ToBase64String(pages[0].ToArray());
                 }
             }
-            return allPages;
         }
 
-        private void SetOptions(HtmlOptions options, string password, int pageNumber)
+        private void CreatePagesInfoFile(string pagesInfoPath, ViewInfo viewInfo)
         {
-            Watermark watermark = null;
-            if (!String.IsNullOrEmpty(globalConfiguration.Viewer.GetWatermarkText()))
-            {
-                // Set watermark properties
-                watermark = new Watermark(globalConfiguration.Viewer.GetWatermarkText());
-                watermark.Color = System.Drawing.Color.Blue;
-                watermark.Position = WatermarkPosition.Diagonal;
-                watermark.Width = 100;
-            }
-            options.EmbedResources = true;
-            // set password for protected document
-            if (!string.IsNullOrEmpty(password))
-            {
-                options.Password = password;
-            }
-            if (watermark != null)
+            var xdoc = new XDocument(new XElement("Pages"));
+
+            foreach (var page in viewInfo.Pages)
             {
-                options.Watermark = watermark;
+                xdoc.Element("Pages")
+                    .Add(new XElement("PageData",
+                        new XElement("Number", page.Number),
+                        new XElement("Angle", 0)));
             }
-            if (pageNumber != 0)
+
+            xdoc.Save(pagesInfoPath);
+        }
+
+        private int GetCurrentAngle(int pageNumber, string pagesInfoPath)
+        {
+            XDocument xdoc = XDocument.Load(pagesInfoPath);
+            var pageData = xdoc.Descendants()?.Elements("Number")?.Where(x => int.Parse(x.Value) == pageNumber)?.Ancestors("PageData");
+            var angle = pageData.Elements("Angle").FirstOrDefault();
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['pageData' is null on at least one execution path.](https://app.codacy.com/manual/GroupDocs/GroupDocs.Viewer-for-.NET-MVC/pullRequest?prid=5242613)